ticket_upgrade_prediction/evaluator.py

```diff
@@ -24,23 +26,72 @@
 from sklearn.model_selection import train_test_split
 
 import mlflow
+from ticket_upgrade_prediction.models import BaseModel
 
 
 @dataclass
 class Metrics:
-    accuracy: float = None
-    roc_auc: float = None
-    precision: float = None
-    recall: float = None
-    f1: float = None
-    pr_auc: float = None
+    accuracy: Optional[float] = None
+    roc_auc: Optional[float] = None
+    precision: Optional[float] = None
+    recall: Optional[float] = None
+    f1: Optional[float] = None
+    pr_auc: Optional[float] = None
+    epoch: Optional[int] = None
 
     def to_dict(self) -> dict:
         return self.__dict__
 
+    @classmethod
+    def from_multiple_metrics(cls, *args):
+        metrics_list = [metric.to_dict() for metric in args]
+        metrics_df = pd.DataFrame(metrics_list).mean()
+        return cls(**metrics_df.to_dict())
+
+    def get_metric_from_string(self, metric_name: str) -> float:
+        """
+        Select one metric from string.
+        Possible metric_name values:
+        accuracy, roc_auc, precision
+        recall, f1, pr_auc, epoch
+        """
+        metric_mapping = {
+            "accuracy": self.accuracy,
+            "roc_auc": self.roc_auc,
+            "precision": self.precision,
+            "recall": self.recall,
+            "f1": self.f1,
+            "pr_auc": self.pr_auc,
+            "epoch": self.epoch,
+        }
+
+        if metric_name not in metric_mapping.keys():
+            raise ValueError(
+                f"Wrong metric name provided! Accepted names: {list(metric_mapping.keys())}"
+            )
+
+        return metric_mapping[metric_name]
+
+
+class BaseEvaluator(ABC):
+    """Interface for Evaluator"""
+
+    @abstractmethod
+    def get_all_metrics(self, to_mlflow: bool):
+        """Method for getting all metrics"""
+
+    @abstractmethod
+    def plot_all_plots(self, save_path: Path, to_mlflow: bool):
+        """Methods for plotting everything"""
 
-class Evaluator:
-    def __init__(self, model: Any, X: pd.DataFrame, y: np.array) -> None:
+
+class Evaluator(BaseEvaluator):
+    def __init__(
+        self,
+        model: BaseModel | ClassifierMixin,
```

Comment:
this line causes an error on my side: `TypeError: unsupported operand type(s) for |: 'ABCMeta' and 'type'`

Reply:
Ancient python moment